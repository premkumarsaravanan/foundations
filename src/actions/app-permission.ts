import { actionCreator } from '../utils/actions'
import ActionTypes from '../constants/action-types'
import { FormState } from '../types/core'
import { ScopeModel } from '@/types/marketplace-api-schema'

export const appPermissionRequestData = actionCreator<string>(ActionTypes.APP_PERMISION_REQUEST_DATA)
export const appPermissionRequestDataFailure = actionCreator<void>(ActionTypes.APP_PERMISION_REQUEST_DATA_FAILURE)
export const appPermissionLoading = actionCreator<boolean>(ActionTypes.APP_PERMISION_LOADING)
export const appPermissionReceiveData = actionCreator<ScopeModel[] | undefined>(ActionTypes.APP_PERMISION_RECEIVE_DATA)
