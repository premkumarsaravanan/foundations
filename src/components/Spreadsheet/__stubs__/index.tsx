import * as React from 'react'
import { Cell, SelectedMatrix } from '../types'
import ReactDataSheet from 'react-datasheet'

export const data: Cell[][] = [
  [
    { readOnly: true, value: 'Office Name' },
    { readOnly: true, value: 'Building Name' },
    { readOnly: true, value: 'Building No.' },
    { readOnly: true, value: 'Address 1' },
    { readOnly: true, value: 'Address 2' },
    { readOnly: true, value: 'Address 3' },
    { readOnly: true, value: 'Address 4' },
    { readOnly: true, value: 'Post Code' },
    { readOnly: true, value: 'Telephone' },
    { readOnly: true, value: 'Fax' },
    { readOnly: true, value: 'Email' }
  ],
  [
    { value: 'London' },
    { value: 'The White House' },
    { value: '15' },
    { value: 'London 1' },
    { value: '' },
    { value: 'Londom 3' },
    { value: '' },
    { value: 'EC12NH' },
    { value: '0845 0000' },
    { value: '' },
    { value: 'row1@gmail.com' }
  ],
  [
    { value: 'London2' },
    { value: 'The Black House' },
    { value: '11' },
    { value: '' },
    { value: '' },
    { value: 'Adress 3' },
    { value: '' },
    { value: 'EC12NH' },
    { value: '087 471 929' },
    { value: '' },
    { value: 'row2@gmail.com' }
  ],
  [
    { value: 'New York' },
    { value: 'Building A' },
    { value: '11' },
    { value: '' },
    { value: '' },
    { value: 'City Z' },
    { value: '' },
    { value: 'AL7187' },
    { value: '017 7162 9121' },
    { value: '' },
    { value: 'row3@gmail.com' }
  ]
]

export const cellRenderProps: ReactDataSheet.CellRendererProps<Cell> = {
  row: 3,
  col: 10,
  cell: { value: 'row3@gmail.com' },
  selected: false,
  editing: false,
  updated: false,
  attributesRenderer: jest.fn() as ReactDataSheet.AttributesRenderer<Cell>,
  className: 'cell',
  style: { background: 'red' },
  onMouseDown: jest.fn(),
  onMouseOver: jest.fn(),
  onDoubleClick: jest.fn(),
  onContextMenu: jest.fn(),
  children: <div>hi</div>
}

export const setData: React.Dispatch<Cell[][]> = jest.fn()

export const setSelected: React.Dispatch<SelectedMatrix> = jest.fn()

export const selectedMatrix = {
  start: { i: 0, j: 1 },
  end: { i: 2, j: 3 }
}
