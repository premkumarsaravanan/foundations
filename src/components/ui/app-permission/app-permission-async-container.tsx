import * as React from 'react'
import { connect } from 'react-redux'
import { ReduxState, FormState } from '@/types/core'
import AsyncContainer from '@/components/ui/async-container'

export const mapStateToProps = (state: ReduxState) => {
  const { loading, error, appPermissionData } = state.appPermission

  return {
    loading,
    error,
    data: appPermissionData
  }
}

const AppPermissionAsyncContainer = connect(mapStateToProps)(AsyncContainer)
AppPermissionAsyncContainer.displayName = 'AppPermissionAsyncContainer'

export default AppPermissionAsyncContainer
