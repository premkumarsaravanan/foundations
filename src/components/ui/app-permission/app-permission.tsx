import * as React from 'react'

import AppPermissionAsyncContainer from './app-permission-async-container'
import AppPermissionWithConnect from './app-permission-content'

const AppPermission = () => {
  return (
    <AppPermissionAsyncContainer>
      <AppPermissionWithConnect />
    </AppPermissionAsyncContainer>
  )
}

export default AppPermission
