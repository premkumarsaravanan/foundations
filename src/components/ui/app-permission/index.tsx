export { default } from './app-permission'
