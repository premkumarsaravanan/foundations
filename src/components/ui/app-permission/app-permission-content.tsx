import * as React from 'react'
import { connect } from 'react-redux'
import { ReduxState } from '@/types/core'
import { ScopeModel } from '@/types/marketplace-api-schema'
import bulma from '@/styles/vendor/bulma'
import appPermissionContentStyles from '@/styles/pages/app-permission-content.scss?mod'
import { Button } from '@reapit/elements'
import { setAppDetailModalStateView, setAppDetailModalStateViewConfirm } from '@/actions/app-detail-modal'

export interface AppPermissionModalMappedProps {
  permissions: ScopeModel[]
  appName: string
}

export interface AppPermissionModalMappedActions {
  setAppDetailModalStateConfirm: () => void
  setAppDetailModalStateView: () => void
}
export type AppPermissionInnerProps = AppPermissionModalMappedProps &
  AppPermissionModalMappedActions & {
    setAppDetailModalStateView: () => void
    afterClose?: () => void
  }

export const handleCloseModal = (afterClose, setAppDetailModalStateView) => () => {
  afterClose()
  setAppDetailModalStateView()
}
export const AppPermissionContent = ({
  permissions,
  setAppDetailModalStateConfirm,
  appName,
  afterClose,
  setAppDetailModalStateView
}: AppPermissionInnerProps) => {
  return (
    <div>
      <h3 className={`${bulma.title} ${bulma.is3} ${appPermissionContentStyles.heading}`}>Permissions Required</h3>
      <h6 className={appPermissionContentStyles.subtitle}>
        {appName} requires the following permissions to integrate with the platform. By installing you are consenting to
        the the following:
      </h6>
      <ul className={appPermissionContentStyles.permissionList}>
        {permissions.map(({ description }) => (
          <li className={appPermissionContentStyles.permissionListItem}>{description}</li>
        ))}
      </ul>
      <div className={appPermissionContentStyles.installButtonContainer}>
        <Button
          dataTest="btnInstall"
          className={appPermissionContentStyles.installButton}
          type="button"
          variant="primary"
          onClick={setAppDetailModalStateConfirm}
        >
          Install
        </Button>
        <Button
          className={appPermissionContentStyles.installButton}
          type="button"
          variant="danger"
          onClick={handleCloseModal(afterClose, setAppDetailModalStateView)}
        >
          Dismiss
        </Button>
      </div>
    </div>
  )
}

export const mapStateToProps = (state: ReduxState): AppPermissionModalMappedProps => ({
  permissions: state.appPermission.appPermissionData || [],
  appName:
    (state.appDetail &&
      state.appDetail.appDetailData &&
      state.appDetail.appDetailData.data &&
      state.appDetail.appDetailData.data.name) ||
    ''
})

export const mapDispatchToProps = (dispatch: any): AppPermissionModalMappedActions => ({
  setAppDetailModalStateView: () => dispatch(setAppDetailModalStateView()),
  setAppDetailModalStateConfirm: () => dispatch(setAppDetailModalStateViewConfirm())
})

const AppPermissionContentInnerWithConnect = connect(
  mapStateToProps,
  mapDispatchToProps
)(AppPermissionContent)

AppPermissionContentInnerWithConnect.displayName = 'AppPermissionContentInnerWithConnect'

export default AppPermissionContentInnerWithConnect
