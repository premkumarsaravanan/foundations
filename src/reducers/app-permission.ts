import { Action, FormState } from '../types/core'
import { isType } from '../utils/actions'
import {
  appPermissionLoading,
  appPermissionReceiveData,
  appPermissionRequestDataFailure
} from '../actions/app-permission'

import { ScopeModel } from '@/types/marketplace-api-schema'

export interface AppPermissionState {
  loading: boolean
  error: boolean
  appPermissionData: ScopeModel[] | null
}

export const defaultState: AppPermissionState = {
  loading: false,
  appPermissionData: null,
  error: false
}

const appPermissionReducer = (state: AppPermissionState = defaultState, action: Action<any>): AppPermissionState => {
  if (isType(action, appPermissionLoading)) {
    return {
      ...state,
      loading: action.data
    }
  }

  if (isType(action, appPermissionReceiveData)) {
    return {
      ...state,
      loading: false,
      error: false,
      appPermissionData: action.data || null
    }
  }

  if (isType(action, appPermissionRequestDataFailure)) {
    return {
      ...state,
      loading: false,
      error: true
    }
  }

  return state
}

export default appPermissionReducer
