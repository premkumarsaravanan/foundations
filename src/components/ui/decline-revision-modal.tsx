import * as React from 'react'
import Modal, { ModalProps } from '@/components/ui/modal'
import { connect } from 'react-redux'
import { ReduxState } from '@/types/core'
import { RejectRevisionModel } from '@/types/marketplace-api-schema'
import { Button, TextArea } from '@reapit/elements'
import { Form, Formik } from 'formik'
import { validate } from '@/utils/form/reject-revision'
import { declineRevision, RevisionDeclineRequestParams } from '@/actions/revision-detail'
import { RevisionDetailState } from '@/reducers/revision-detail'
import { oc } from 'ts-optchain'
import CallToAction from './call-to-action'

interface DeclineRevisionInnerWithConnectOwnProps {
  closeModal?: () => void
  onDeclineSuccess: () => void
  visible: boolean
}

export interface DeclineRevisionModalMappedProps extends DeclineRevisionInnerWithConnectOwnProps {
  revisionDetail: RevisionDetailState
  name: string
  email: string
}

export interface DeclineRevisionModalMappedActions {
  submitDeclineRevision: (params: RevisionDeclineRequestParams) => void
}

const mapStateToProps = (
  state: ReduxState,
  ownProps: DeclineRevisionInnerWithConnectOwnProps
): DeclineRevisionModalMappedProps => ({
  revisionDetail: state.revisionDetail,
  email: state.auth.loginSession!.loginIdentity.email,
  name: state.auth.loginSession!.loginIdentity.name,
  closeModal: ownProps.closeModal,
  onDeclineSuccess: ownProps.onDeclineSuccess,
  visible: ownProps.visible
})

const mapDispatchToProps = (dispatch: any): DeclineRevisionModalMappedActions => ({
  submitDeclineRevision: params => dispatch(declineRevision(params))
})

export type DeclineRevisionModalProps = Pick<ModalProps, 'visible' | 'afterClose'> &
  DeclineRevisionModalMappedProps &
  DeclineRevisionModalMappedActions

export const DeclineRevisionModal: React.FunctionComponent<DeclineRevisionModalProps> = ({
  visible,
  submitDeclineRevision,
  afterClose,
  revisionDetail,
  onDeclineSuccess,
  name,
  email
}) => {
  const { declineFormState } = revisionDetail
  const { appId, id: appRevisionId } = oc<RevisionDetailState>(revisionDetail).revisionDetailData.data({})
  const [rejectionReason, setRejectionReason] = React.useState('')

  const isLoading = declineFormState === 'SUBMITTING'
  const isSuccessed = declineFormState === 'SUCCESS'

  return (
    <Modal
      visible={visible}
      size="small"
      afterClose={() => {
        if (isSuccessed) {
          onDeclineSuccess()
        } else if (!isLoading && afterClose) {
          afterClose()
        }
      }}
    >
      <Formik
        initialValues={{ email, name, rejectionReason } as RejectRevisionModel}
        data-test="revision-decline-form"
        validate={validate}
        onSubmit={(formValues: RejectRevisionModel) => {
          if (appId && appRevisionId) {
            setRejectionReason(formValues.rejectionReason || '')
            submitDeclineRevision({ appId, appRevisionId, ...formValues })
          }
        }}
        render={() => {
          return isSuccessed ? (
            <CallToAction
              title="Rejected!"
              buttonText="Back to List"
              dataTest="decline-revision-success-message"
              onButtonClick={() => {
                onDeclineSuccess()
              }}
              isCenter
            >
              Revision has been declined successfully.
            </CallToAction>
          ) : (
            <Form>
              <TextArea
                name="rejectionReason"
                id="rejectionReason"
                labelText="Rejection reason"
                dataTest="revision-rejection-reason"
              />
              <div className="flex justify-end">
                <Button
                  type="button"
                  variant="secondary"
                  className="mr-2"
                  disabled={Boolean(isLoading)}
                  onClick={() => afterClose && afterClose()}
                  dataTest="revision-decline-cancel"
                >
                  Cancel
                </Button>
                <Button
                  type="submit"
                  variant="danger"
                  loading={Boolean(isLoading)}
                  disabled={Boolean(isLoading)}
                  dataTest="revision-decline-submit"
                >
                  Decline
                </Button>
              </div>
            </Form>
          )
        }}
      />
    </Modal>
  )
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DeclineRevisionModal)
