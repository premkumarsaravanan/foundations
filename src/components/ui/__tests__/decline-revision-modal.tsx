import * as React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import { DeclineRevisionModal, DeclineRevisionModalProps } from '../decline-revision-modal'
import { appPermissionStub } from '@/sagas/__stubs__/app-permission'

const props: DeclineRevisionModalProps = {
  onDeclineSuccess: jest.fn(),
  email: 'email',
  name: 'name',
  submitDeclineRevision: jest.fn(),
  visible: true,
  // @ts-ignore: only pick necessary props to test
  revisionDetail: {
    revisionDetailData: {
      data: {
        appId: 'appIDd',
        id: 'revisionID'
      },
      scopes: appPermissionStub
    },
    approveFormState: 'PENDING'
  }
}

describe('DeclineRevisionModal', () => {
  it('should match a snapshot', () => {
    expect(toJson(shallow(<DeclineRevisionModal {...props} />))).toMatchSnapshot()
  })

  afterEach(() => {
    jest.resetAllMocks()
  })
})
