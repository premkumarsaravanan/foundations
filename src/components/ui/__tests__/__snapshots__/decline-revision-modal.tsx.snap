// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeclineRevisionModal should match a snapshot 1`] = `
<Modal
  afterClose={[Function]}
  size="small"
  visible={true}
>
  <Formik
    enableReinitialize={false}
    initialValues={
      Object {
        "email": "email",
        "name": "name",
        "rejectionReason": "",
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    render={[Function]}
    validate={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  />
</Modal>
`;
